docs/design-proposals/2023-05 vParquet3.md

```diff
@@ -0,0 +1,179 @@
+---
+Authors: Adrian Stoewer (@stoewer), Mario Rodriguez (@mapno)
+Created: May 2023
+Last updated: 2023-05-31
+---
+
+# vParquet3
+
+## Summary
+
+This document describes how to extend Tempo’s Parquet based block format such that a configurable subset of span attributes can be stored in dedicated columns.
+The goal is to improve the search performance of TraceQL in Tempo.
+
+### Table of Contents
+
+* [Context](#context)
+* [Goals and requirements](#goals-and-requirements)
+* [Implementation](#implementation)
+  * [Schema with spare columns](#schema-with-spare-columns)
+  * [Configuration and overrides](#configuration-and-overrides)
+  * [Compaction](#compaction)
+  * [Tooling](#tooling)
+
+## Context
+The current Parquet schema stores span attributes in the repeated group `Attrs`.
+The group contains a single column for all attribute keys and separate columns for the corresponding values (depending on the value type).
+The size of the span attribute columns combined contribute to 30-70% of the overall block size.
+To search for a specific attribute, it is necessary to access this large chunk of data.
+The analysis of Tempo blocks from different use cases has further revealed that the 10 most frequently used span attributes are responsible for over 50% of the overall attribute size.
+
+A benchmark with a schema that stores the 5 most frequently occurring attributes in dedicated columns showed significant performance improvements:
+
+* The block size decreased by ~10%
+* Compressed size of the generic attribute columns was reduced by ~40%
+* The search speed for queries without matches improved by ~80% for attributes with dedicated columns and ~70% for generic attributes columns
+* The search speed for queries returning matches improved by ~50% for attributes with dedicated columns and ~30% for generic attribute columns
+
+The amount and kind of span attributes varies a lot between use cases and the above benchmark was conducted with tracing data that captured a lot of span attributes.
+Therefore, the search speed improvements that can be achieved by storing span attributes in dedicated columns,
+can deviate from the above results.
+However, the data still shows that dedicated columns for span attributes that can be tuned to a specific use case can improve Tempo’s search performance significantly.
+This document describes how dedicated columns for span attributes will be implemented in Tempo.
+
+## Goals and requirements
+
+There are different ways dedicated columns for attributes can be added to Tempo's block format and how they are configured and tuned for the specific workload.
```

Comment:
this sentence is confusing to me. i think you could just drop it and the point is clear.

Reply:
Agree, dropped.